src/tiered/tiered_handle.c

```diff
@@ -791,13 +791,41 @@ __wt_tiered_open(WT_SESSION_IMPL *session, const char *cfg[])
     return (ret);
 }
 
+/*
+ * __tiered_cleanup --
+ *     Cleanup a tiered data handle.
+ */
+static void
+__tiered_cleanup(WT_SESSION_IMPL *session, WT_TIERED *tiered)
+{
+    uint32_t i;
+
+    /* Cleanup and dereference all tiers.  */
+    for (i = 0; i < WT_TIERED_MAX_TIERS; i++) {
+        if (tiered->tiers[i].tier != NULL && tiered->tiers[i].tier->session_inuse > 0)
+            (void)__wt_atomic_subi32(&tiered->tiers[i].tier->session_inuse, 1);
```

Comment:
@sueloverso @ddanderson I'm not 100% sure whether I need to decrement `session_inuse` here.

Reply:
It seems that we do need to dereference it since we increment it in `__tiered_dhandle_setup`:
https://github.com/wiredtiger/wiredtiger/blob/b017f238da1ffd71a0623f6bc2e9eb2425887d44/src/tiered/tiered_handle.c#L109

Reply:
I think we should use the function `__wt_cursor_dhandle_decr_use`, you may have to wrap the call using `WT_WITH_DHANDLE`.  And all the calls to increment this counter in this file should use `__wt_cursor_dhandle_incr_use`.

And also, these lines of code to dereference are identical to the lines in tiered cleanup.  Maybe refactor to get rid of that duplication?

Reply:
@sueloverso my code was actually wrong because during connection close `__wt_conn_dhandle_discard` may discard tier's dhandle and we trying to access memory after free. For this reason, I made changes to pass through the `final` flag. If it's true we must not try to access the tier pointer.

Reply:
@ddanderson good suggestion, thank you! I'll change the code.